File: sdk/go/client/lifecycle.go
package client

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"sync"
	"sync/atomic"
	"syscall"
	"time"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/protocol"
)

// Config is the worker's static configuration.
type Config struct {
	WorkerCode     string
	TenantID       string
	BuildID        string
	SDKVersion     string
	CapabilityTags []string
	// RegisterAttributes is the fingerprint attribute map sent at register time.
	// It is scanned by SensitiveValidator (§1.8) before send.
	RegisterAttributes map[string]any
	MaxConcurrentTasks int
	// HeartbeatInterval / LeaseRenewInterval default to the §5 values when zero.
	HeartbeatInterval  time.Duration
	LeaseRenewInterval time.Duration
	// StopTimeout is the graceful-stop budget (default 30s, §1.6).
	StopTimeout time.Duration
}

func (c Config) withDefaults() Config {
	if c.MaxConcurrentTasks <= 0 {
		c.MaxConcurrentTasks = 1
	}
	if c.HeartbeatInterval <= 0 {
		c.HeartbeatInterval = DefaultHeartbeatInterval
	}
	if c.LeaseRenewInterval <= 0 {
		c.LeaseRenewInterval = DefaultLeaseRenewInterval
	}
	if c.StopTimeout <= 0 {
		c.StopTimeout = 30 * time.Second
	}
	return c
}

// Worker wires the runtime engine: FSM + transport + consumer + schedulers +
// handler. Start() registers, subscribes, and spins the consume + scheduler
// loops; Stop(timeout) runs the §1.6 graceful-stop sequence.
type Worker struct {
	cfg       Config
	transport Transport
	consumer  Consumer
	handler   TaskHandler
	validator *SensitiveValidator
	fsm       *FSM
	registry  *InFlightRegistry
	pipeline  *MessagePipeline
	logger    *log.Logger

	heartbeat *HeartbeatScheduler
	lease     *LeaseRenewalScheduler

	// rootCtx governs all loops; rootCancel wakes them on Stop.
	rootCtx    context.Context
	rootCancel context.CancelFunc

	wg       sync.WaitGroup // consume + scheduler goroutines
	taskWg   sync.WaitGroup // in-flight handler executions
	draining atomic.Bool
	stopOnce sync.Once
	started  atomic.Bool
}

// NewWorker builds a worker. consumer + handler are required; validator
// defaults to a protocol-keyword validator when nil.
func NewWorker(cfg Config, transport Transport, consumer Consumer, handler TaskHandler, validator *SensitiveValidator, logger *log.Logger) *Worker {
	cfg = cfg.withDefaults()
	if logger == nil {
		logger = log.Default()
	}
	if validator == nil {
		validator = NewSensitiveValidator()
	}
	fsm := NewFSM()
	registry := NewInFlightRegistry()
	w := &Worker{
		cfg:       cfg,
		transport: transport,
		consumer:  consumer,
		handler:   handler,
		validator: validator,
		fsm:       fsm,
		registry:  registry,
		logger:    logger,
	}
	w.pipeline = NewMessagePipeline(cfg.TenantID, fsm, cfg.MaxConcurrentTasks, registry.Count, logger)
	w.heartbeat = NewHeartbeatScheduler(transport, fsm, cfg.WorkerCode, cfg.TenantID,
		WithHeartbeatInterval(cfg.HeartbeatInterval),
		WithHeartbeatInFlight(registry.Count),
		WithHeartbeatLogger(logger),
		WithOnDrain(w.beginDrain),
	)
	w.lease = NewLeaseRenewalScheduler(transport, registry, cfg.WorkerCode, cfg.TenantID,
		WithLeaseInterval(cfg.LeaseRenewInterval),
		WithLeaseLogger(logger),
	)
	return w
}

// FSM exposes the runtime FSM (tests / observability).
func (w *Worker) FSM() *FSM { return w.fsm }

// InFlight returns the current in-flight task count.
func (w *Worker) InFlight() int { return w.registry.Count() }

// Heartbeat exposes the heartbeat scheduler (tests).
func (w *Worker) Heartbeat() *HeartbeatScheduler { return w.heartbeat }

// Start registers the worker, applies the register-online decision, and spins
// the consume loop + both schedulers. Returns a register error (fatal auth /
// sensitive-data rejection) without starting loops.
func (w *Worker) Start(ctx context.Context) error {
	req := RegisterRequest{
		WorkerCode:     w.cfg.WorkerCode,
		TenantID:       w.cfg.TenantID,
		BuildID:        w.cfg.BuildID,
		SDKVersion:     w.cfg.SDKVersion,
		CapabilityTags: w.cfg.CapabilityTags,
		Attributes:     w.cfg.RegisterAttributes,
		// #536 register-time protocol-version gate: advertise the SDK's current
		// major (last of SupportedSchemaVersions) so the platform identifies us.
		ProtocolVersion: protocol.SupportedSchemaVersions[len(protocol.SupportedSchemaVersions)-1],
	}
	// §1.8 register-path sensitive scan (fail-fast at startup).
	if err := w.validator.ValidateRegister(req); err != nil {
		return err
	}
	res, err := w.transport.Register(ctx, req)
	if err != nil {
		return err
	}

	// Apply register-online decision: FSM -> NORMAL, start schedulers, subscribe.
	d := protocol.DecideRegister(res.Idempotent)
	if d.FsmTransition != nil {
		w.fsm.SetState(WorkerState(*d.FsmTransition))
	}
	w.logger.Printf("INFO worker registered code=%s idempotent=%v state=%s",
		w.cfg.WorkerCode, res.Idempotent, w.fsm.State())

	w.rootCtx, w.rootCancel = context.WithCancel(ctx)
	w.started.Store(true)

	w.wg.Add(3)
	go func() { defer w.wg.Done(); w.heartbeat.Run(w.rootCtx) }()
	go func() { defer w.wg.Done(); w.lease.Run(w.rootCtx) }()
	go func() { defer w.wg.Done(); w.consumeLoop(w.rootCtx) }()
	return nil
}

// consumeLoop polls the consumer and dispatches accepted records to handlers.
// It honors draining (rejects new work) and the FSM Kafka pause flag.
func (w *Worker) consumeLoop(ctx context.Context) {
	for {
		if ctx.Err() != nil {
			return
		}
		if w.draining.Load() {
			return // §1.6: stop accepting new messages.
		}
		records, ok := w.consumer.Poll()
		if !ok {
			return // woken / closed.
		}
		for _, rec := range records {
			if w.draining.Load() || ctx.Err() != nil {
				return
			}
			msg, disp := w.pipeline.OnMessage(rec)
			switch disp {
			case DispositionAccepted:
				w.dispatch(ctx, msg)
				w.consumer.Commit()
			case DispositionBackpressure:
				// partition paused; leave offset uncommitted, retry later.
				w.consumer.Pause()
			case DispositionRejectedSchema, DispositionDroppedForeignTenant, DispositionDecodeError:
				// do NOT commit offset (§1.9 / §A); drop and move on.
			}
		}
	}
}

// dispatch claims, validates parameters, runs the handler, and reports the
// terminal result. Runs the handler in a goroutine bounded by maxConcurrent via
// the in-flight registry (backpressure gate already enforced upstream).
func (w *Worker) dispatch(ctx context.Context, msg TaskDispatchMessage) {
	claim, err := w.transport.Claim(ctx, msg.TaskID, msg.IdempotencyKey, ClaimRequest{
		TenantID: w.cfg.TenantID,
		WorkerID: w.cfg.WorkerCode,
	})
	if err != nil {
		if _, ok := errAsDrop(err); ok {
			w.logger.Printf("WARN claim dropped taskId=%s: %v", msg.TaskID, err)
			return
		}
		w.logger.Printf("WARN claim failed taskId=%s: %v", msg.TaskID, err)
		return
	}
	if claim.Idempotent {
		// 409: already claimed by another worker; treat as no-op success (§B).
		w.logger.Printf("INFO claim idempotent (already claimed) taskId=%s", msg.TaskID)
		return
	}

	sig := NewCancellationSignal(ctx)
	w.registry.Add(msg.TaskID, sig)
	w.taskWg.Add(1)
	go func() {
		defer w.taskWg.Done()
		defer w.registry.Remove(msg.TaskID)
		defer func() {
			// resume partition once a slot frees if we were paused.
			if w.fsm.Paused() && w.registry.Count() <= w.cfg.MaxConcurrentTasks {
				w.fsm.SetPaused(false)
				w.consumer.Resume()
			}
		}()
		// A panicking handler must not crash the worker process: recover, log,
		// and report the failure as a terminal result. An SdkTaskStopped that
		// propagated all the way out (decision 三: cooperative cancel after a
		// committed safe point) is NOT a failure — it maps to a CANCELLED
		// terminal report. The handler is documented to let SdkTaskStopped
		// propagate (return it / panic with it), never swallow it.
		defer func() {
			if r := recover(); r != nil {
				if stop, ok := r.(*SdkTaskStopped); ok {
					w.logger.Printf("INFO task stopped (cooperative cancel) taskId=%s breakPosition=%v",
						msg.TaskID, stop.BreakPosition)
					w.report(msg, stoppedResult(stop))
					return
				}
				w.logger.Printf("ERROR handler panic taskId=%s: %v", msg.TaskID, r)
				w.report(msg, Fail(protocol.ErrorCodeExecutionFailed, fmt.Sprintf("panic: %v", r)))
			}
		}()

		// §1.8 parameters-path sensitive scan -> SECURITY_REJECTED.
		if res, rejected := w.validator.ValidateParameters(claim.EffectiveConfig); rejected {
			w.report(msg, res)
			return
		}

		tc := &TaskContext{
			TaskID:          msg.TaskID,
			EffectiveConfig: claim.EffectiveConfig,
			TraceID:         claim.TraceID,
			Cancellation:    sig,
			Progress:        NoopProgressReporter{},
		}
		result := w.handler.Execute(tc)
		// If cancelled and handler didn't set a terminal cancel code, normalize.
		if sig.IsCancellationRequested() && result.IsSuccess() {
			result = Fail(protocol.ErrorCodeCancelled, "task cancelled")
		}
		w.report(msg, result)
	}()
}

// stoppedResult maps an SdkTaskStopped sentinel to a CANCELLED terminal result
// (decision 三). The committed break position is surfaced in outputs so the
// platform / a later resume can see where the task safely stopped.
func stoppedResult(stop *SdkTaskStopped) TaskResult {
	r := Fail(protocol.ErrorCodeCancelled, "task stopped at committed safe point")
	if stop != nil && stop.BreakPosition != nil {
		r.Outputs = map[string]any{"breakPosition": stop.BreakPosition}
	}
	return r
}

// ResultFromError maps a handler error to a terminal TaskResult. SdkTaskStopped
// becomes a CANCELLED report (not a failure); any other error becomes
// EXECUTION_FAILED. Handlers that prefer returning a TaskResult (rather than
// letting SdkTaskStopped panic-propagate) can do `return ResultFromError(err)`.
func ResultFromError(err error) TaskResult {
	if err == nil {
		return Success(nil, "")
	}
	if stop, ok := IsSdkTaskStopped(err); ok {
		return stoppedResult(stop)
	}
	return Fail(protocol.ErrorCodeExecutionFailed, err.Error())
}

// report posts the terminal result with the dispatch idempotency key. It uses a
// fresh background context (NOT rootCtx) so the terminal report still lands
// during graceful Stop, which cancels rootCtx (§1.6 — a cancelled report would
// otherwise lose the task outcome).
func (w *Worker) report(msg TaskDispatchMessage, result TaskResult) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	req := NewReportRequest(msg.TaskID, w.cfg.TenantID, w.cfg.WorkerCode, result)
	if err := w.transport.Report(ctx, msg.TaskID, msg.IdempotencyKey, req); err != nil {
		w.logger.Printf("WARN report failed taskId=%s code=%s: %v", msg.TaskID, result.ErrorCode, err)
	}
}

// errAsDrop reports whether the error is a 404 (drop-and-forget).
func errAsDrop(err error) (*NotFoundError, bool) {
	var nf *NotFoundError
	if asNotFound(err, &nf) {
		return nf, true
	}
	return nil, false
}

// beginDrain flips the worker into DRAINING and triggers Stop with the
// configured budget (invoked from the heartbeat DRAINING directive, §1.5).
func (w *Worker) beginDrain() {
	w.fsm.SetState(StateDraining)
	go w.Stop(w.cfg.StopTimeout)
}

// Stop runs the §1.6 graceful-stop sequence exactly once:
//
//	draining=true (reject new) -> consumer.Wakeup() -> drain in-flight (timeout*0.4)
//	-> executor shutdown wait (timeout*0.6) -> Deactivate.
//
// PlanStop captures the plan (kafka=wakeup, deactivate, drainThenDeactivate).
func (w *Worker) Stop(timeout time.Duration) {
	w.stopOnce.Do(func() {
		if !w.started.Load() {
			return
		}
		plan := protocol.PlanStop(int(timeout / time.Millisecond))
		w.logger.Printf("INFO graceful stop worker=%s within=%dms", w.cfg.WorkerCode, *plan.WithinMs)

		w.draining.Store(true) // reject new Kafka messages immediately.
		w.fsm.SetState(StateDraining)

		// wake the consumer poll so consumeLoop exits.
		w.consumer.Wakeup()

		// stop schedulers + consume loop.
		if w.rootCancel != nil {
			w.rootCancel()
		}

		drainBudget := time.Duration(float64(timeout) * 0.4)
		execBudget := time.Duration(float64(timeout) * 0.6)

		// phase 1: wait for in-flight to drain within 40% of the budget.
		w.waitInFlight(drainBudget)
		// phase 2: wait for handler goroutines to finish within 60%.
		w.waitTasks(execBudget)
		// scheduler/consume goroutines are bounded too; don't block forever.
		w.wg.Wait()

		w.consumer.Close()

		// §1.6 final step: say goodbye. Best-effort with a short bounded ctx.
		dctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if err := w.transport.Deactivate(dctx, w.cfg.WorkerCode); err != nil {
			w.logger.Printf("WARN deactivate failed worker=%s: %v", w.cfg.WorkerCode, err)
		} else {
			w.logger.Printf("INFO worker deactivated code=%s", w.cfg.WorkerCode)
		}
	})
}

// waitInFlight blocks until the in-flight count hits zero or the budget expires.
func (w *Worker) waitInFlight(budget time.Duration) {
	deadline := time.Now().Add(budget)
	for w.registry.Count() > 0 && time.Now().Before(deadline) {
		time.Sleep(10 * time.Millisecond)
	}
}

// waitTasks waits for handler goroutines within the budget.
func (w *Worker) waitTasks(budget time.Duration) {
	done := make(chan struct{})
	go func() { w.taskWg.Wait(); close(done) }()
	select {
	case <-done:
	case <-time.After(budget):
		w.logger.Printf("WARN executor shutdown timed out worker=%s, %d task(s) still running",
			w.cfg.WorkerCode, w.registry.Count())
	}
}

// RunUntilSignal starts the worker and blocks until SIGINT/SIGTERM, then runs
// Stop(30s) (byo-sdk-guide §1.6 K8s terminationGracePeriodSeconds=30s).
func (w *Worker) RunUntilSignal(ctx context.Context) error {
	sigCtx, stop := signal.NotifyContext(ctx, os.Interrupt, syscall.SIGTERM)
	defer stop()
	if err := w.Start(sigCtx); err != nil {
		return err
	}
	<-sigCtx.Done()
	w.logger.Printf("INFO signal received worker=%s -> Stop(30s)", w.cfg.WorkerCode)
	w.Stop(30 * time.Second)
	return nil
}

File: sdk/typescript/src/client/lifecycle.ts
/**
 * WorkerLifecycle (§1.5 FSM + §1.6 graceful stop).
 *
 * 4-state FSM (NORMAL / DEGRADED / PAUSED / DRAINING). `start()` registers,
 * starts the schedulers, then subscribes the consumer. `stop(timeoutMs)` sets
 * DRAINING, wakes the consumer, drains in-flight within timeout*0.4, awaits the
 * executor within timeout*0.6, then deactivates (wire-protocol §4 stop order).
 *
 * §4 Node pits handled:
 *   - a global `unhandledRejection` handler so a stray promise rejection cannot
 *     hang SIGTERM.
 *   - SIGTERM → stop(30000) (K8s terminationGracePeriodSeconds default).
 */

import {
  decideBackpressure,
  decideRegister,
  newIdempotencyKey,
  planStop,
} from "../decide.ts";
import type { FsmState, KafkaAction } from "../protocol.ts";
import { FatalTransportError, type ReportBody, type Transport } from "./transport.ts";
import {
  HeartbeatScheduler,
  LeaseRenewalScheduler,
  type InFlightTask,
} from "./scheduler.ts";
import {
  MessagePipeline,
  assignmentOf,
  consoleLogger,
  type Consumer,
  type ConsumerRecord,
  type DispatchMessage,
  type Logger,
} from "./consumer.ts";
import {
  SimpleCancellationSignal,
  NoopProgressReporter,
  type TaskContext,
  type TaskHandler,
  type TaskResult,
} from "./handler.ts";
import {
  InMemorySdkCheckpoint,
  ResumeSupport,
  SdkTaskStopped,
  type SdkCheckpoint,
  type ResumeOptions,
} from "./checkpoint.ts";
import { SensitiveDataValidator } from "./sensitive.ts";
import { ErrorCode } from "../protocol.ts";
import { SUPPORTED_SCHEMA_VERSIONS } from "../constants.ts";

let unhandledRejectionInstalled = false;

/** Install a process-wide unhandledRejection guard exactly once (§4 pit). */
export function installUnhandledRejectionGuard(logger: Logger = consoleLogger): void {
  if (unhandledRejectionInstalled) return;
  unhandledRejectionInstalled = true;
  process.on("unhandledRejection", (reason) => {
    logger.error("unhandledRejection swallowed (would otherwise hang SIGTERM)", {
      reason: String(reason),
    });
  });
}

export interface WorkerConfig {
  tenantId: string;
  workerCode: string;
  maxConcurrent: number;
  /**
   * workerTypes this worker serves (== the taskTypes its handler accepts). A
   * dispatch message whose `workerType` is not served is not-for-us and is not
   * committed. Omitted / empty → serve all (no routing filter).
   */
  workerTypes?: readonly string[];
  registerBody?: Record<string, unknown>;
  buildHeartbeatBody?: () => Record<string, unknown>;
  buildRenewBody?: (taskId: string) => Record<string, unknown>;
}

export interface WorkerLifecycleDeps {
  config: WorkerConfig;
  transport: Transport;
  consumer: Consumer;
  handler: TaskHandler;
  heartbeatIntervalMs?: number;
  leaseRenewIntervalMs?: number;
  logger?: Logger;
  validator?: SensitiveDataValidator;
  /** test seam: avoid registering a real SIGTERM listener. */
  installSignalHandlers?: boolean;
  /**
   * ADR-037 §决策一 — per-task break-point store factory. The real tenant impl
   * persists checkpoints in the same transaction as its business data (see
   * `SdkCheckpoint`). Defaults to an in-memory store (tests / examples only).
   */
  checkpointFactory?: (taskId: string) => SdkCheckpoint;
  /** ADR-037 §决策二 — progress-throttle / self-report tuning for `ctx.commit`. */
  resumeOptions?: ResumeOptions;
}

interface TrackedTask extends InFlightTask {
  promise: Promise<void>;
  /** ADR-014 invocation token from the claim response; echoed on renew + report. */
  partitionInvocationId?: string | null;
}

export class WorkerLifecycle {
  #cfg: WorkerConfig;
  #transport: Transport;
  #consumer: Consumer;
  #handler: TaskHandler;
  #logger: Logger;
  #validator: SensitiveDataValidator;
  #checkpointFactory: (taskId: string) => SdkCheckpoint;
  #resumeOptions?: ResumeOptions;

  #fsm: FsmState = "NORMAL";
  #draining = false;
  #inFlight = new Map<string, TrackedTask>();
  #heartbeat: HeartbeatScheduler;
  #leaseRenewal: LeaseRenewalScheduler;
  #pipeline: MessagePipeline;
  #sigtermHandler?: () => void;

  constructor(deps: WorkerLifecycleDeps) {
    this.#cfg = deps.config;
    this.#transport = deps.transport;
    this.#consumer = deps.consumer;
    this.#handler = deps.handler;
    this.#logger = deps.logger ?? consoleLogger;
    this.#validator = deps.validator ?? new SensitiveDataValidator();
    this.#checkpointFactory =
      deps.checkpointFactory ?? (() => new InMemorySdkCheckpoint());
    this.#resumeOptions = deps.resumeOptions;

    this.#heartbeat = new HeartbeatScheduler(
      this.#transport,
      {
        workerCode: this.#cfg.workerCode,
        // WorkerHeartbeatDto requires [tenantId, workerCode, status, heartbeatAt].
        buildBody:
          this.#cfg.buildHeartbeatBody ??
          (() => ({
            tenantId: this.#cfg.tenantId,
            workerCode: this.#cfg.workerCode,
            status: this.#draining ? "DRAINING" : "RUNNING",
            heartbeatAt: new Date().toISOString(),
            currentLoad: this.#inFlight.size,
          })),
        setFsm: (s) => this.#setFsm(s),
        applyKafka: (a) => this.#applyKafka(a),
        onDrain: () => void this.stop(30_000),
        setMaxConcurrent: (n) => {
          this.#cfg.maxConcurrent = n;
        },
        logger: this.#logger,
      },
      deps.heartbeatIntervalMs ?? 30_000,
    );

    this.#leaseRenewal = new LeaseRenewalScheduler(
      this.#transport,
      {
        inFlight: () => [...this.#inFlight.values()],
        // TaskClaimRequest (claim/renew shared) requires [tenantId, workerId];
        // taskId is in the URL. Echo the ADR-014 invocation token when present.
        buildBody:
          this.#cfg.buildRenewBody ??
          ((taskId) => ({
            tenantId: this.#cfg.tenantId,
            workerId: this.#cfg.workerCode,
            partitionInvocationId:
              this.#inFlight.get(taskId)?.partitionInvocationId ?? null,
          })),
        dropTask: (taskId) => this.#inFlight.delete(taskId),
        logger: this.#logger,
      },
      deps.leaseRenewIntervalMs ?? 60_000,
    );

    this.#pipeline = new MessagePipeline({
      tenantId: this.#cfg.tenantId,
      workerTypes: this.#cfg.workerTypes,
      inFlight: () => this.#inFlight.size,
      maxConcurrent: this.#cfg.maxConcurrent,
      assignment: assignmentOf(this.#consumer),
      logger: this.#logger,
      onAccepted: (msg) => this.#handleAccepted(msg),
    });

    if (deps.installSignalHandlers !== false) {
      installUnhandledRejectionGuard(this.#logger);
      this.#sigtermHandler = () => {
        this.#logger.info("SIGTERM received; graceful stop(30000)");
        void this.stop(30_000);
      };
      process.on("SIGTERM", this.#sigtermHandler);
    }
  }

  get fsm(): FsmState {
    return this.#fsm;
  }

  get isDraining(): boolean {
    return this.#draining;
  }

  get inFlightCount(): number {
    return this.#inFlight.size;
  }

  #setFsm(state: FsmState): void {
    if (this.#fsm === state) return;
    this.#logger.info("FSM transition", { from: this.#fsm, to: state });
    this.#fsm = state;
    if (state === "DRAINING") this.#draining = true;
  }

  #applyKafka(action: KafkaAction): void {
    switch (action) {
      case "pause":
        this.#consumer.pause();
        break;
      case "resume":
        this.#consumer.resume();
        break;
      case "subscribe":
      case "wakeup":
      case "none":
      case "drop-message":
        break;
    }
  }

  /** start(): register → start schedulers → subscribe consumer (§4 order). */
  async start(): Promise<void> {
    // register body credential check (§1.8 register path → throw on leak).
    // WorkerHeartbeatDto requires [tenantId, workerCode, status, heartbeatAt].
    const regBody = this.#cfg.registerBody ?? {
      tenantId: this.#cfg.tenantId,
      workerCode: this.#cfg.workerCode,
      // ADR-035 §2: SDK self-hosted workers register under the fixed "sdk-self-hosted"
      // group. The platform requires it (worker_registry.worker_group is NOT NULL) and
      // selects SDK workers by it; omitting it makes register fail with HTTP 500. The
      // decision-core (decide.ts) already encodes this value; the default body lacked it.
      workerGroup: "sdk-self-hosted",
      status: "RUNNING",
      heartbeatAt: new Date().toISOString(),
      // #536 register-time protocol-version gate: advertise the SDK's current
      // major (last of SUPPORTED_SCHEMA_VERSIONS). Register only — heartbeat null.
      protocolVersion:
        SUPPORTED_SCHEMA_VERSIONS[SUPPORTED_SCHEMA_VERSIONS.length - 1],
    };
    this.#validator.assertRegisterBody(regBody);

    const ack = await this.#transport.register(regBody);
    const decision = decideRegister(ack.idempotent);
    this.#setFsm(decision.fsmTransition);
    this.#logger.info("registered", { idempotent: ack.idempotent });

    this.#heartbeat.start();
    this.#leaseRenewal.start();

    // subscribe last so no message arrives before schedulers are live
    await this.#consumer.start((r: ConsumerRecord) =>
      this.#onRecord(r),
    );
  }

  async #onRecord(record: ConsumerRecord): Promise<void> {
    if (this.#draining) {
      // draining: refuse new messages (do not commit so a fresh worker re-reads)
      return;
    }
    await this.#pipeline.onMessage(record);
  }

  /** claim → execute → report for one accepted, in-tenant message. */
  async #handleAccepted(msg: DispatchMessage): Promise<void> {
    const cancellation = new SimpleCancellationSignal();
    // fixture 24: mint a FRESH key per distinct write — never a fixed claim-{id}.
    const idemKey = newIdempotencyKey();

    const run = (async (): Promise<void> => {
      try {
        // §1.8 task-params path: SECURITY_REJECTED instead of throwing
        const paramScan = this.#validator.scanTaskParams(msg.parameters);
        if (!paramScan.ok) {
          await this.#report(msg.taskId, {
            success: false,
            errorCode: ErrorCode.SECURITY_REJECTED,
            resultSummary: `credential in parameters: ${paramScan.leakedKeys.join(", ")}`,
          });
          return;
        }

        const claim = await this.#transport.claim(msg.taskId, idemKey);
        // record the ADR-014 invocation token so renew/report can echo it
        const tracked = this.#inFlight.get(msg.taskId);
        if (tracked) tracked.partitionInvocationId = claim.partitionInvocationId ?? null;
        const progress = new NoopProgressReporter();
        const resume = new ResumeSupport({
          taskId: msg.taskId,
          checkpoint: this.#checkpointFactory(msg.taskId),
          progress,
          cancellation,
          options: this.#resumeOptions,
        });
        const ctx: TaskContext = {
          taskId: msg.taskId,
          effectiveConfig: claim.effectiveConfig ?? {},
          traceId:
            claim.traceId ??
            (msg.runtimeAttributes?.traceId as string | undefined) ??
            "",
          cancellation,
          progress,
          checkpoint: () => resume.checkpoint(),
          commit: (breakPosition) => resume.commit(breakPosition),
        };

        let result: TaskResult;
        try {
          result = await this.#handler.execute(ctx);
        } catch (e) {
          // ADR-037 §决策三 — cooperative cancel lands here as a *cancelled*
          // terminal report, not a failure.
          if (e instanceof SdkTaskStopped) {
            result = {
              success: false,
              errorCode: ErrorCode.CANCELLED,
              resultSummary: "task stopped at checkpoint (cancelled)",
              outputs: { breakPosition: e.breakPosition },
            };
          } else {
            result = {
              success: false,
              errorCode: ErrorCode.EXECUTION_FAILED,
              resultSummary: String(e),
            };
          }
        }

        await this.#report(msg.taskId, {
          success: result.success,
          errorCode: result.errorCode,
          outputs: result.outputs,
          resultSummary: result.resultSummary,
          partitionInvocationId:
            this.#inFlight.get(msg.taskId)?.partitionInvocationId ?? null,
        });
      } catch (e) {
        // claim/report fatal — log, then fail-fast on auth errors (§ openapi:
        // 401/403 must stop the worker; a stale/invalid credential won't fix itself).
        this.#logger.error("task pipeline failed", {
          taskId: msg.taskId,
          error: String(e),
        });
        if (e instanceof FatalTransportError && (e.status === 401 || e.status === 403)) {
          this.#logger.error("auth failure (401/403); failing fast", {
            taskId: msg.taskId,
            status: e.status,
          });
          void this.stop(30_000);
        }
      } finally {
        this.#inFlight.delete(msg.taskId);
        // A slot just freed. Route the resume decision through the shared core so
        // the max/2 hysteresis applies: stay paused while in-flight is still in
        // the [max/2, max) band, resume only once it drops below max/2. This
        // avoids pause/resume thrash at the capacity boundary.
        if (this.#consumer.isPaused() && !this.#draining) {
          const bp = decideBackpressure(
            this.#inFlight.size,
            this.#cfg.maxConcurrent,
            true,
          );
          if (bp.action === "backpressure" && bp.kafka === "resume") {
            this.#consumer.resume();
          }
        }
      }
    })();

    this.#inFlight.set(msg.taskId, {
      taskId: msg.taskId,
      cancellation,
      promise: run,
    });
  }

  async #report(taskId: string, body: ReportBody): Promise<void> {
    // fixture 24: fresh key, never a fixed report-{taskId} (a redelivered task's
    // report would otherwise replay the platform's stale first outcome).
    await this.#transport.report(taskId, body, newIdempotencyKey());
  }

  /**
   * stop(timeoutMs): DRAINING → wakeup → drain in-flight (40%) → await executor
   * (60%) → deactivate. Order per wire-protocol §4.
   */
  async stop(timeoutMs = 30_000): Promise<void> {
    if (this.#draining && this.#fsm === "DRAINING" && this.#inFlight.size === 0) {
      // already stopping with nothing in flight — still ensure deactivate ran
    }
    const plan = planStop(timeoutMs);
    this.#draining = true;
    this.#setFsm("DRAINING");

    // wake the consumer (interrupt poll); refuse new messages
    if (plan.kafka === "wakeup") {
      this.#consumer.pause();
      await this.#consumer.wakeup();
    }

    // Keep heartbeat + lease renewal running THROUGH drain: a worker that stops
    // heartbeating during the (up to timeoutMs) drain window can be flagged dead
    // by the platform's missed-heartbeat reaper and have its still-running
    // in-flight tasks redispatched (double-run). Stop both AFTER drain.

    // phase 1: drain in-flight within 40% of the window
    const drainDeadline = Date.now() + timeoutMs * 0.4;
    await this.#awaitInFlight(drainDeadline);

    // phase 2: await executor (remaining tasks' promises) within 60%
    const execDeadline = Date.now() + timeoutMs * 0.6;
    await this.#awaitInFlight(execDeadline);

    this.#heartbeat.stop();
    this.#leaseRenewal.stop();

    // deactivate last (§4)
    if (plan.deactivate) {
      try {
        await this.#transport.deactivate(this.#cfg.workerCode);
        this.#logger.info("deactivated");
      } catch (e) {
        this.#logger.warn("deactivate failed (heartbeat-stop fallback applies)", {
          error: String(e),
        });
      }
    }

    if (this.#sigtermHandler) {
      process.off("SIGTERM", this.#sigtermHandler);
      this.#sigtermHandler = undefined;
    }
  }

  async #awaitInFlight(deadline: number): Promise<void> {
    while (this.#inFlight.size > 0 && Date.now() < deadline) {
      const promises = [...this.#inFlight.values()].map((t) => t.promise);
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await Promise.race([
        Promise.allSettled(promises),
        new Promise((r) => setTimeout(r, Math.min(remaining, 25)).unref?.()),
      ]);
    }
  }
}
